ngraph/src/ngraph/runtime/reference/scatter_update.hpp

```diff
@@ -0,0 +1,115 @@
+//*****************************************************************************
+// Copyright 2017-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <cstring>
+
+#include "ngraph/check.hpp"
+#include "ngraph/coordinate_transform.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            template <typename DataType, typename IndicesType>
+            void scatter_update(const DataType* input_data,
+                                const IndicesType* indices,
+                                const DataType* updates,
+                                const int64_t& axis,
+                                DataType* out_buf,
+                                const Shape& data_shape,
+                                const Shape& indices_shape,
+                                const Shape& updates_shape)
+            {
+                // Copy inputs to out
+                std::memcpy(out_buf, input_data, sizeof(DataType) * shape_size(data_shape));
+
+                // Algorithm overview
+                // data[..., indices[m, n, ..., p], ...] = updates[..., m, n, ..., p, ...]
+                // where first ... in the data corresponds to first axis dimensions,
+                // last ... in the data corresponds to the rank(data) - (axis + 1) dimensions.
+
+                //
+                // for i_coord in indices[m, n, ..., p]:
+                //     i_idx = index(i_coord)
+                //     for d_coord in slice data[..., i_idx, ...] &&
```

Comment:
This comment makes understanding the algorithm harder. What does "slice data" mean? And how should we understand two for-loops joined with `&&`?

Reply:
Slice here has NumPy meaning so, `data[..., i_idx, ....]` is a shortcut for a _slice_ of n-dimensional tensor of input data. The ellipsis stands for `:` symbol, which gets all values from respective axis, for all dimensions up to `i_idx` (the axis at which we have update indexes) and following it till the end.